### DISCUSSION TITLE
Export to cloud storage providers?

### DISCUSSION THREAD

**Discussion by RhetTbull** [RhetTbull/osxphotos#658] [General]
It occurred to me today as I was thinking about a better way to backup my own Photos library that it would be possible, with relatively minor changes, to create a plugin for osxphotos that exports directly to S3, backblaze, or other cloud providers.  That is, osxphotos would directly upload to the cloud instead of exporting to the local file system.  Most of the functions today reside in the FileUtil class in osxphotos/fileutil.py which is how osxphotos supports `--dry-run` mode.  There are a couple other methods that would need to be added to FileUtil but then one could conceivably subclass FileUtil with a version that uses S3 or some other cloud storage provider.  

I wonder if others would be interested in such a feature?

**Comment by oPromessa** [comment]
Yup. That'd would be neat!

I'll tell you my story... while working with Local vs Cloud and free Cloud storage. 

### Local vs Cloud
Haven't come around to decide which Storage cloud server to best use to buy. So for now I trust the NAS Server and my backups to local disks... etc. It's still the cheaper option per Terabyte.
* To access from the Web I use the `osxphotos` exported files on the NAS with [Synology Photos](https://www.synology.com/en-global/DSM70/SynologyPhotos)

### Cloud (free services)
* My approach in the past, while Flickr provided 1 Terabyte free storage with an API, was to:
   * first export from iPhoto (at the time) with [Phoshare](https://github.com/lhagan/phoshare) into a NAS Server 
   * and then with [flick-uploader](https://github.com/oPromessa/flickr-uploader) upload it into Flickr.
* Then Flickr was no longer available and I started using [Synology sync with Amazon Drive](https://kb.synology.com/en-us/DSM/tutorial/Discontinued_access_to_Amazon_Drive) to benefit from unlimited storage for Prime users (for Pics).
   * that was also then cancelled.
* Nowadays I use `osxphotos` into the NAS or a local exported folder in the MAC and then I use [Amazon Photos Desktop](https://www.amazon.com/gp/help/customer/display.html?nodeId=GQFQXL85XKLKS6QE) to upload... 
   * which does not cater very for syncing... works mostly for uploading
   * And with the Prime account you get free storage for Pics but not for Video.
* The Synology NAS (once it has the files) can backup via their [Hyper Backup Tool](https://www.synology.com/en-uk/dsm/feature/hyper_backup) to several Cloud Storage servers.
   * <img width="505" alt="image" src="https://user-images.githubusercontent.com/21261491/157520063-8d926584-6c03-40ed-a53f-33e8dbffea03.png">

**Reply by RhetTbull** [reply]
For users with a NAS your workflow seems to be good because you get both a local copy and a cloud copy if you have the NAS do the synch to cloud.  For those without a setup like yours, I think an advantage of this proposed idea is you could backup a large Photos library to your cloud provider without needing to have enough storage space on the local machine or NAS to export the entire library.  If I design the plug-in interface right, people could develop a plug-in for the cloud provider of choice that "just works" with osxphotos.
